R/obic_recommendations.R

```diff
@@ -206,4 +206,162 @@ obic_recommendations <- function(dt.recom) {
+  dt.score <- copy(dt.score)
+  dt.score[,ID := .I]
+  
+  # Add B_LU_BRP ad B_SOILTYPE_AGR to dt.score
+  dt.score[, B_LU_BRP := B_LU_BRP]
+  dt.score[, B_SOILTYPE_AGR := B_SOILTYPE_AGR]
+  
+  # add local databases for joining properties -----
+  
+  # load db for measures, crop and soil categories
+  m.obi_bkp <- OBIC::recom.obic_bkp
+  crops.obic <- as.data.table(OBIC::crops.obic)
+  soils.obic <- as.data.table(OBIC::soils.obic)
+  
+  # merge dt.score with soils.obic and crops.obic
+  dt.score <- merge(dt.score, OBIC::soils.obic[, list(soiltype, soiltype.m)], by.x = "B_SOILTYPE_AGR", by.y = "soiltype")
+  dt.score <- merge(dt.score, OBIC::crops.obic[, list(crop_code, crop_measure)], by.x = "B_LU_BRP", by.y = "crop_code")
+  
+  # redesign measures db to make it suitable for joining -----
+  
+  # drop indicators from measure db that are not present in dt.score
+  mdb1 <- m.obi_bkp[indicator %in% colnames(dt.score)]
+  
+  # drop descriptive columns that are not needed
+  mdb1[,c('m_description','m_soilfunction') := NULL]
+  
+  # check if there is missing data for any indicator
+  cols <- colnames(dt.score)[grepl('^I_C_|^I_B_|^I_P_',colnames(dt.score))]
+  cols <- cols[!(grepl('^I_P|^I_B',cols) & grepl('_BCS$',cols))]
+  cols <- cols[!cols %in% unique(mdb1$indicator)]
+  checkmate::assert_character(cols, len = 0)
+  
+  # setkeys
+  setkey(mdb1,indicator,m_sector,m_soiltype)
+  
+  # Make a new datatable to store the effects of measures -----
+  
+  # add local copy of the input
+  dt.recom <- copy(dt.score) 
+  
+  # melt the database
+  cols_m <- colnames(dt.recom)[grepl('^I_C_|^I_P_|^I_B_',colnames(dt.recom))]
+  cols_i <- colnames(dt.recom)[grepl('^B_|^ID$|soil|crop',colnames(dt.recom))]
+  dt.recom <- melt(dt.recom,id.vars = cols_i,measure.vars = cols_m,variable.name = 'indicator', value.name='score')
+  
+  # add weighing factor based on number of indicators
+  dt.recom[,cat := tstrsplit(indicator,'_',keep = 2)]
+  
+  # Determine amount of indicators per category
+  dt.recom.ncat <- dt.recom[,list(ncat = .N),by = list(ID, cat)]
+  
+  # add number of indicators per category
+  dt.recom <- merge(dt.recom,dt.recom.ncat,by=c("ID", "cat"),all.x = TRUE)
+  
+  # calculate weighing factor depending on number of indicators
+  dt.recom[,cf := log(ncat + 1)]
+  
+  # reset names and set key
+  setnames(dt.recom,c('crop_measure','soiltype.m'),c('m_sector','m_soiltype'))
+  setkey(dt.recom,indicator,m_sector,m_soiltype)
+  
+  # join measures and dt.score ------
+  
+  # join each parcel with possible measures (join by sector, soil type and indicator)
+  dt.recom2 <- mdb1[dt.recom,allow.cartesian=TRUE]
+  
+  # add parameter whether score is below threshold yes or no
+  dt.recom2[, threshold := fifelse(score <= m_threshold & score >= 0,1,0)]
+  
+  # calculate the score of the measure for each indicator
+  dt.recom2[, score.m := cf * m_applicability * (pmax(0,m_effect) * threshold + pmin(0,m_effect))]
```

Comment:
Just to be sure: is it `pmin` or `pmax`? Now it looks like positive values are not allowed

Reply:
This is correct. If there is positive effect and the threshold is reached, the effect is accounted for a result of the pmax. If the effect is negative, the effect is set to 0, because with the pmin the negative value is selected. Also if there is a negative effect and the threshold is not reached, a negative score will be calculated because pmax selects 0 and as a result of the pmin the negative value will be used for calculating score.m